File: 2025/src/days/day_01.rs
pub fn solve(input: &str) -> (String, String) {
    (part1(input), part2(input))
}

fn part1(input: &str) -> String {
    let lines: Vec<&str> = input.lines().collect();

    let mut pos: i64 = 50;
    let mut count: u64 = 0;

    for line in lines {
        let dir = &line[0..1];
        let dist = &line[1..];
        let abs_dist = dist.parse::<i64>().unwrap() % 100;

        match dir {
            "L" => pos -= abs_dist,
            "R" => pos += abs_dist,
            _ => panic!("Invalid direction {}", dir),
        }

        if pos < 0 {
            pos = 100 + pos;
        } else if pos > 99 {
            pos = pos - 100;
        }

        if pos == 0 {
            count += 1
        }
    }

    return count.to_string();
}

fn part2(input: &str) -> String {
    let lines: Vec<&str> = input.lines().collect();

    let mut pos: i64 = 50;
    let mut count: u64 = 0;

    for line in lines {
        let direction: i64 = if line.chars().next().unwrap() == 'L' {
            -1
        } else {
            1
        };
        let magnitude: i64 = line[1..].parse().unwrap();

        let old_position = pos;
        let new_pos = pos + direction * magnitude;

        let q = new_pos.div_euclid(100);
        let r = new_pos.rem_euclid(100);

        count += q.abs() as u64 + if direction == -1 && r == 0 { 1 } else { 0 }
            - if direction == -1 && old_position == 0 {
                1
            } else {
                0
            };

        pos = r;
    }

    return count.to_string();
}

File: 2025/day01.py
def part_one(lines):
    pos = 50
    count = 0

    for line in lines:
        dir = line[0]
        step = int(line[1:])
        abs_dist = step % 100

        pos += abs_dist if dir == "R" else -abs_dist

        if pos % 100 == 0:
            count += 1

    print(f"Part 1 password: {count}")


def part_two(lines):
    pos = 50
    count = 0

    for line in lines:
        dir = line[0]
        step = int(line[1:])

        for _ in range(step):
            pos += 1 if dir == "R" else -1

            if pos % 100 == 0:
                count += 1

    print(f"Part 2 password: {count}")


def main(lines):
    part_one(lines)
    part_two(lines)
